### PR TITLE
ci: cross python wheel

### PR DESCRIPTION
**Pull request by xslingcn**
We can build a single python wheel for all supported python versions.

- Remove the python versions matrix
- Add `--py-limited-api` config in building commands

### CODE REVIEW

**Comment by xslingcn** [comment]
Unable to test it now because the self-hosted runner is not set up in github actions at the moment.
[https://github.com/flashinfer-ai/flashinfer/blob/7627a986ca3f799c45f452e82a11115449271363/.github/workflows/release_wheel.yml#L36C1-L36C27](url)

**Comment by xslingcn** [comment]
Can't test it now because the self-hosted runner is not set up in github actions at the moment.

https://github.com/flashinfer-ai/flashinfer/blob/56f8eca41f01b819d082b3adde13c62e10d4305e/.github/workflows/release_wheel.yml#L37-L41

**Comment by yzh119** [comment]
Hi @youkaichao  @comaniac , we are using `--py-limited-api` to build wheels for more python versions, do you think we did it correctly?

**Comment by xslingcn** [comment]
> Hi @youkaichao @comaniac , we are using `--py-limited-api` to build wheels for more python versions, do you think we did it correctly?

The usage of Limited API is not very well documented for setuptools, but I've found some references:
https://github.com/pypa/packaging.python.org/issues/1376
https://github.com/pytorch/pytorch/pull/138088

**Comment by youkaichao** [comment]
Looks reasonable. can you test it?

File: .github/workflows/release_wheel.yml
Comment (yzh119) [inline .github/workflows/release_wheel.yml:1]: I suppose we no longer need
```yaml

      - uses: softprops/action-gh-release@v1
        with:
          tag_name: ${{ inputs.tag_name }}
          files: |
            dist/flashinfer*cp38*.whl
      - uses: softprops/action-gh-release@v1
        with:
          tag_name: ${{ inputs.tag_name }}
          files: |
            dist/flashinfer*cp39*.whl
      - uses: softprops/action-gh-release@v1
        with:
          tag_name: ${{ inputs.tag_name }}
          files: |
            dist/flashinfer*cp310*.whl
      - uses: softprops/action-gh-release@v1
        with:
          tag_name: ${{ inputs.tag_name }}
          files: |
            dist/flashinfer*cp311*.whl
      - uses: softprops/action-gh-release@v1
        with:
          tag_name: ${{ inputs.tag_name }}
          files: |
            dist/flashinfer*cp312*.whl
```
We can change them to
```yaml
      - uses: softprops/action-gh-release@v1
        with:
          tag_name: ${{ inputs.tag_name }}
          files: |
            dist/flashinfer*.whl
```

Comment (xslingcn) [reply .github/workflows/release_wheel.yml:1]: yeah makes sense

**Review by yzh119 — APPROVED** [review]
Let's first check whether this works on nightly wheels.

**Comment by zhyncs** [comment]
Hi @xslingcn 

```
[2024-12-16 07:09:10 TP0] TpModelWorkerClient hit an exception: Traceback (most recent call last):
  File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker_overlap_thread.py", line 103, in forward_thread_func
    self.forward_thread_func_()
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker_overlap_thread.py", line 134, in forward_thread_func_
    logits_output, next_token_ids = self.worker.forward_batch_generation(
  File "/sgl-workspace/sglang/python/sglang/srt/managers/tp_worker.py", line 154, in forward_batch_generation
    logits_output = self.model_runner.forward(forward_batch)
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 693, in forward
    return self.forward_extend(forward_batch)
  File "/sgl-workspace/sglang/python/sglang/srt/model_executor/model_runner.py", line 659, in forward_extend
    self.attn_backend.init_forward_metadata(forward_batch)
  File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/flashinfer_backend.py", line 151, in init_forward_metadata
    self.indices_updater_prefill.update(
  File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/flashinfer_backend.py", line 552, in update_single_wrapper
    self.call_begin_forward(
  File "/sgl-workspace/sglang/python/sglang/srt/layers/attention/flashinfer_backend.py", line 686, in call_begin_forward
    wrapper_paged.begin_forward(
  File "/usr/local/lib/python3.10/dist-packages/flashinfer/prefill.py", line 1440, in plan
    self._cached_module = get_batch_prefill_sm90_module(*get_module_args)
  File "/usr/local/lib/python3.10/dist-packages/flashinfer/prefill.py", line 224, in get_batch_prefill_sm90_module
    from . import _kernels_sm90
ImportError: Python version mismatch: module was compiled for Python 3.11, but the interpreter version is incompatible: 3.10.16 (main, Dec  4 2024, 08:53:37) [GCC 9.4.0].
```
It doesn't seem to work as expected.

**Comment by zhyncs** [comment]
ref https://github.com/flashinfer-ai/flashinfer-nightly/releases/tag/0.1.6%2B51236c9

**Comment by zhyncs** [comment]
https://github.com/flashinfer-ai/flashinfer-nightly/releases/tag/0.1.6%2B6819a0f
`6819a0f5d975cbb11c121cfffe6289f4629bab2f` works

```
pip3 install https://github.com/flashinfer-ai/flashinfer-nightly/releases/download/0.1.6%2B6819a0f/flashinfer-0.1.6+6819a0f.cu121torch2.4-cp310-cp310-linux_x86_64.whl --force-reinstall --no-deps
```

**Comment by yzh119** [comment]
The reason that cross-python wheel is broken is because pybind uses some APIs that is not part of Py_LIMITED_API : https://github.com/pybind/pybind11/issues/1755 (thanks @youkaichao for informing me about this).

We should use TORCH_LIBRARY instead: https://github.com/pytorch/pytorch/blob/2293fe1024812d6349f6e2b3b7de82c6b73f11e4/torch/library.h#L934-L943

cc @xslingcn @zhyncs

**Pull request merged**